source/ranges.tex

```diff
@@ -602,54 +680,72 @@
 
 \rSec2[range.prim.size]{\tcode{ranges::size}}
 \pnum
-The name \tcode{size} denotes a customization point
-object\iref{customization.point.object}. The expression
-\tcode{ranges::size(E)} for a subexpression \tcode{E} of type
-\tcode{T} is expression-equivalent to:
+The name \tcode{ranges::size} denotes a customization point
+object\iref{customization.point.object}.
+
+\pnum
+Given a subexpression \tcode{E} with type \tcode{T},
+let \tcode{t} be an lvalue that
+denotes the same object as \tcode{E} if \tcode{E} is a glvalue, and
+otherwise denotes the result of applying
+the temporary materialization conversion\iref{conv.rval} to \tcode{E}.
+Then:
+
 \begin{itemize}
 \item
-  \tcode{\placeholdernc{decay-copy}(extent_v<T>)} if \tcode{T} is an array
-  type\iref{basic.compound}.
+  If \tcode{T} is an array of unknown bound\iref{dcl.array},
+  \tcode{ranges::size(E)} is ill-formed.
 
 \item
-  Otherwise, if
-  \tcode{disable_sized_range<remove_cv_t<T>>}\iref{range.sized}
-  is \tcode{false}:
-  \begin{itemize}
-  \item
-    \tcode{\placeholdernc{decay-copy}(E.size())}
-    if it is a valid expression and its type \tcode{I}
-    is integer-like\iref{iterator.concept.winc}.
+  Otherwise, if \tcode{T} is an array type,
+  \tcode{ranges::size(E)} is expression-equivalent to
+  \tcode{\placeholdernc{decay-copy}(extent_v<T>)}.
 
-  \item
-    Otherwise, \tcode{\placeholdernc{decay-copy}(size(E))}
-    if it is a valid expression and its type \tcode{I}
-    is integer-like
-    with overload resolution performed in a context that includes
-    the declaration:
+\item
+  Otherwise, if \tcode{disable_sized_range<remove_cv_t<T>>}\iref{range.sized}
+  is \tcode{false} and
+  \tcode{\placeholdernc{decay-copy}(t.size())}
+  is a valid expression of integer-like type\iref{iterator.concept.winc},
+  \tcode{ranges::size(E)} is expression-equivalent to
+  \tcode{\placeholdernc{decay-copy}(E.size())}.
+
+\item
+  Otherwise, if \tcode{T} is a class or enumeration type,
+  \tcode{disable_sized_range<remove_cv_t<T>>} is \tcode{false} and
+  \tcode{\placeholdernc{decay-copy}(size(t))}
+  is a valid expression of integer-like type
+  with overload resolution performed in a context in which
+  unqualified lookup for \tcode{size} finds only the declarations
 \begin{codeblock}
-template<class T> void size(T&&) = delete;
+void size(auto&) = delete;
+void size(const auto&) = delete;
 \end{codeblock}
-    and does not include a declaration of \tcode{ranges::size}.
-  \end{itemize}
+  then \tcode{ranges::size(E)} is expression-equivalent to
+  \tcode{\placeholdernc{decay-copy}(size(E))}
```

Comment:
Ditto typo.
```suggestion
  \tcode{\placeholdernc{decay-copy}(size(t))}
```

Reply:
I think we still have the same typo here in `ranges::data`:

https://github.com/cplusplus/draft/blob/caf7428fd6c807f30538042cf9bf1b27b8772b01/source/ranges.tex#L918-L922
We check that `ranges::begin(t)` is valid, and then do `ranges::begin(E)` instead. It should have the same effects, but it's inconsistent and confusing to switch between `t` and `E`.

I've proposed a fix in #4628